server/server.go

```diff
@@ -2682,28 +2687,62 @@ func (s *Server) ProfilerAddr() *net.TCPAddr {
 	return s.profiler.Addr().(*net.TCPAddr)
 }
 
-// ReadyForConnections returns `true` if the server is ready to accept clients
-// and, if routing is enabled, route connections. If after the duration
-// `dur` the server is still not ready, returns `false`.
-func (s *Server) ReadyForConnections(dur time.Duration) bool {
+func (s *Server) readyForConnections(d time.Duration) error {
 	// Snapshot server options.
 	opts := s.getOpts()
 
-	end := time.Now().Add(dur)
+	type info struct {
+		ok  bool
+		err error
+	}
+	chk := make(map[string]info)
+
+	end := time.Now().Add(d)
 	for time.Now().Before(end) {
 		s.mu.Lock()
-		ok := s.listener != nil &&
-			(opts.Cluster.Port == 0 || s.routeListener != nil) &&
-			(opts.Gateway.Name == "" || s.gatewayListener != nil) &&
-			(opts.LeafNode.Port == 0 || s.leafNodeListener != nil) &&
-			(opts.Websocket.Port == 0 || s.websocket.listener != nil)
+		chk["server"] = info{ok: s.listener != nil, err: s.listenerErr}
+		chk["route"] = info{ok: (opts.Cluster.Port == 0 || s.routeListener != nil), err: s.routeListenerErr}
+		chk["gateway"] = info{ok: (opts.Gateway.Name == "" || s.gatewayListener != nil), err: s.gatewayListenerErr}
+		chk["leafNode"] = info{ok: (opts.LeafNode.Port == 0 || s.leafNodeListener != nil), err: s.leafNodeListenerErr}
+		chk["websocket"] = info{ok: (opts.Websocket.Port == 0 || s.websocket.listener != nil), err: s.websocket.listenerErr}
```

Comment:
I realize that we don't have that for mqtt, and can't think of a reason right now why. Would you mind adding a s.mqtt.listenerErr and all that?

Reply:
Riiight! I also noticed that actually! Great idea.